File: src/main/java/eladkay/quaritum/common/block/base/BlockModColored.java
package eladkay.quaritum.common.block.base;

import net.minecraft.block.material.MapColor;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyEnum;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.EnumDyeColor;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.relauncher.Side;

import static eladkay.quaritum.common.item.chalk.ItemChalk.capitalizeFirst;

public class BlockModColored extends BlockMod {
    public static final PropertyEnum<EnumDyeColor> COLOR = PropertyEnum.create("color", EnumDyeColor.class);

    public BlockModColored(String name, Material materialIn) {
        super(name, materialIn, (String[]) (FMLCommonHandler.instance().getEffectiveSide() == Side.CLIENT ? generateVariants(name) : null));
        this.setDefaultState(this.blockState.getBaseState().withProperty(COLOR, EnumDyeColor.WHITE));
        this.setCreativeTab(CreativeTabs.BUILDING_BLOCKS);
    }

    public static String[] generateVariants(String name) {
        String[] ret = new String[16];
        for (EnumDyeColor dye : EnumDyeColor.values())
            ret[dye.ordinal()] = name + capitalizeFirst(dye.toString());
        return ret;
    }

    @Override
    public int damageDropped(IBlockState state) {
        return state.getValue(COLOR).getMetadata();
    }

    @Override
    public MapColor getMapColor(IBlockState state) {
        return state.getValue(COLOR).getMapColor();
    }

    public IBlockState getStateFromMeta(int meta) {
        return this.getDefaultState().withProperty(COLOR, EnumDyeColor.byMetadata(meta));
    }

    @Override
    public int getMetaFromState(IBlockState state) {
        return state.getValue(COLOR).getMetadata();
    }

    @Override
    protected BlockStateContainer createBlockState() {
        return new BlockStateContainer(this, COLOR);
    }

}

File: src/main/java/eladkay/quaritum/common/block/base/BlockModColored.kt
package eladkay.quaritum.common.block.base

import com.teamwizardry.librarianlib.common.base.block.BlockMod
import eladkay.quaritum.common.lib.capitalizeFirst
import net.minecraft.block.material.MapColor
import net.minecraft.block.material.Material
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.EnumDyeColor

open class BlockModColored(name: String, materialIn: Material) : BlockMod(name, materialIn, *BlockModColored.generateVariants(name)) {

    init {
        this.defaultState = this.blockState.baseState.withProperty(COLOR, EnumDyeColor.WHITE)
        this.setCreativeTab(CreativeTabs.BUILDING_BLOCKS)
    }

    override fun damageDropped(state: IBlockState?): Int {
        return state!!.getValue(COLOR).metadata
    }

    override fun getMapColor(state: IBlockState?): MapColor {
        return state!!.getValue(COLOR).mapColor
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        return this.defaultState.withProperty(COLOR, EnumDyeColor.byMetadata(meta))
    }

    override fun getMetaFromState(state: IBlockState?): Int {
        return state!!.getValue(COLOR).metadata
    }

    override fun createBlockState(): BlockStateContainer {
        return BlockStateContainer(this, COLOR)
    }

    companion object {
        val COLOR = PropertyEnum.create<EnumDyeColor>("color", EnumDyeColor::class.java)

        fun generateVariants(name: String): Array<String> {
            val ret = arrayOf<String>()
            for (dye in EnumDyeColor.values())
                ret[dye.ordinal] = name + dye.toString().capitalizeFirst()
            return ret
        }
    }

}
